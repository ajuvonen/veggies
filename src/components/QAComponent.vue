<script lang="ts" setup>
import {Disclosure, DisclosureButton, DisclosurePanel} from '@headlessui/vue';
import IconComponent from './IconComponent.vue';

const questions = [
  'appPurpose',
  'studyDetails',
  'whatCounts',
  'categories',
  'isItFree',
  'isMyDataSafe',
  'contact',
];

const keysWithLinks = ['contact', 'studyDetails'];
</script>
<template>
  <ContentElement :title="$t('qa.title')" class="min-h-0" labelTag="label">
    <div class="flex-container flex-col has-scroll">
      <Disclosure v-for="key in questions" :key="key" v-slot="{open}">
        <DisclosureButton class="QA__button" :data-test-id="`qa-button-${key}`">
          <span>{{ $t(`qa.${key}.title`) }}</span>
          <IconComponent :class="open ? 'rotate-180 transform' : ''" icon="chevron" />
        </DisclosureButton>
        <transition
          enter-active-class="transition duration-100 ease-out"
          enter-from-class="transform h-0 opacity-0"
          enter-to-class="transform h-full opacity-100"
          leave-active-class="transition duration-75 ease-out"
          leave-from-class="transform h-full opacity-100"
          leave-to-class="transform h-0 opacity-0"
        >
          <DisclosurePanel class="QA__panel" :data-test-id="`qa-panel-${key}`">
            <i18n-t :keypath="`qa.${key}.text`" tag="div" scope="global">
              <a
                v-if="keysWithLinks.includes(key)"
                :href="$t(`qa.${key}.link.url`)"
                target="_blank"
                noreferrer
                noopener
                >{{ $t(`qa.${key}.link.text`) }}</a
              >
            </i18n-t>
          </DisclosurePanel>
        </transition>
      </Disclosure>
    </div>
  </ContentElement>
</template>
<style lang="scss" scoped>
.QA__button {
  @apply w-full button-like;
  @apply flex-container justify-between;
  @apply bg-sky-500;

  &:hover,
  &:focus {
    @apply bg-sky-600;
  }
}

.QA__panel {
  @apply p-2;
}
</style>
