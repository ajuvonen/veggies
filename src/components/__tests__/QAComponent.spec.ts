import {describe, it, expect} from 'vitest';
import {mount} from '@vue/test-utils';
import QAComponent from '@/components/QAComponent.vue';
import {nextTick} from 'vue';

describe('QAComponent', () => {
  it('renders', () => {
    const wrapper = mount(QAComponent);
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.findAll('[aria-expanded="false"]').length).toBe(7);
  });

  it('renders answers', async () => {
    const wrapper = mount(QAComponent);
    wrapper.findAll('.QA__button').forEach(async (button) => {
      await button.trigger('click');
    });
    await nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.findAll('[aria-expanded="false"]').length).toBe(0);
  });
});
