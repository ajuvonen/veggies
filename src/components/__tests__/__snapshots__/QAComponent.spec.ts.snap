// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QAComponent > renders 1`] = `
"<div data-v-bcc04aab="" class="flex-container flex-col min-h-0"><label class="label-like">Common Questions</label>
  <div data-v-bcc04aab="" class="flex-container flex-col has-scroll"><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-0" type="button" aria-expanded="false" data-headlessui-state="" class="QA__button" data-test-id="qa-button-appPurpose"><span data-v-bcc04aab="">What's this app for?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <!---->
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-2" type="button" aria-expanded="false" data-headlessui-state="" class="QA__button" data-test-id="qa-button-studyDetails"><span data-v-bcc04aab="">What's the science?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <!---->
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-4" type="button" aria-expanded="false" data-headlessui-state="" class="QA__button" data-test-id="qa-button-whatCounts"><span data-v-bcc04aab="">What counts as a portion?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <!---->
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-6" type="button" aria-expanded="false" data-headlessui-state="" class="QA__button" data-test-id="qa-button-categories"><span data-v-bcc04aab="">But strawberry is a...</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <!---->
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-8" type="button" aria-expanded="false" data-headlessui-state="" class="QA__button" data-test-id="qa-button-isItFree"><span data-v-bcc04aab="">Is the app monetized?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <!---->
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-10" type="button" aria-expanded="false" data-headlessui-state="" class="QA__button" data-test-id="qa-button-isMyDataSafe"><span data-v-bcc04aab="">Is my data safe?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <!---->
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-12" type="button" aria-expanded="false" data-headlessui-state="" class="QA__button" data-test-id="qa-button-contact"><span data-v-bcc04aab="">How to get in touch?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <!---->
    </transition-stub>
  </div>
</div>"
`;

exports[`QAComponent > renders answers 1`] = `
"<div data-v-bcc04aab="" class="flex-container flex-col min-h-0"><label class="label-like">Common Questions</label>
  <div data-v-bcc04aab="" class="flex-container flex-col has-scroll"><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-0" type="button" aria-expanded="true" data-headlessui-state="open" class="QA__button" data-test-id="qa-button-appPurpose" aria-controls="headlessui-disclosure-panel-v-1"><span data-v-bcc04aab="">What's this app for?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0 rotate-180 transform" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <div data-v-bcc04aab="" id="headlessui-disclosure-panel-v-1" data-headlessui-state="open" class="QA__panel" data-test-id="qa-panel-appPurpose">
        <div data-v-bcc04aab="">Eat Your Veggies is a free app for tracking plant-based ingredients on weekly basis in your diet. According to a study, consuming 30 or more veggies weekly has multiple health benefits over a less versatile diet. The key, according to study, is not as much in the amounts you consume, but in variation.</div>
      </div>
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-2" type="button" aria-expanded="true" data-headlessui-state="open" class="QA__button" data-test-id="qa-button-studyDetails" aria-controls="headlessui-disclosure-panel-v-3"><span data-v-bcc04aab="">What's the science?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0 rotate-180 transform" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <div data-v-bcc04aab="" id="headlessui-disclosure-panel-v-3" data-headlessui-state="open" class="QA__panel" data-test-id="qa-panel-studyDetails">
        <div data-v-bcc04aab="">I'm neither a nutritian nor a biologist, so I don't want to make exaggerated claims or interpretations. I'm not affiliated with the study in any way, but if you want to read more about the study, look up The American Gut Project from the journal of <a data-v-bcc04aab="" href="https://doi.org/10.1128/msystems.00031-18" target="_blank" noreferrer="" noopener="">American Society for Microbiology (ASM)</a>.</div>
      </div>
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-4" type="button" aria-expanded="true" data-headlessui-state="open" class="QA__button" data-test-id="qa-button-whatCounts" aria-controls="headlessui-disclosure-panel-v-5"><span data-v-bcc04aab="">What counts as a portion?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0 rotate-180 transform" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <div data-v-bcc04aab="" id="headlessui-disclosure-panel-v-5" data-headlessui-state="open" class="QA__panel" data-test-id="qa-panel-whatCounts">
        <div data-v-bcc04aab="">Ultimately, how you use the app is up to your discretion. However, consider that minuscule amounts might not be of any nutritional value. Therefore I personally consider only portions of a heaping tablespoon or larger adequate. It is also worth noting that very processed foods don't count towards a portion. So ketchup should not equal a tomato.</div>
      </div>
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-6" type="button" aria-expanded="true" data-headlessui-state="open" class="QA__button" data-test-id="qa-button-categories" aria-controls="headlessui-disclosure-panel-v-7"><span data-v-bcc04aab="">But strawberry is a...</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0 rotate-180 transform" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <div data-v-bcc04aab="" id="headlessui-disclosure-panel-v-7" data-headlessui-state="open" class="QA__panel" data-test-id="qa-panel-categories">
        <div data-v-bcc04aab="">...false fruit, I know. Also tomato is not a vegetable in biological sense, so the app takes some liberties in its categories. Still, please report any major inaccuracies you come across! You can also request missing veggies to be added. They should be such that are commercially available globally.</div>
      </div>
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-8" type="button" aria-expanded="true" data-headlessui-state="open" class="QA__button" data-test-id="qa-button-isItFree" aria-controls="headlessui-disclosure-panel-v-9"><span data-v-bcc04aab="">Is the app monetized?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0 rotate-180 transform" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <div data-v-bcc04aab="" id="headlessui-disclosure-panel-v-9" data-headlessui-state="open" class="QA__panel" data-test-id="qa-panel-isItFree">
        <div data-v-bcc04aab="">The app is free to use, and it has no advertisements. There will be a possibility to donate, but don't feel obliged.</div>
      </div>
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-10" type="button" aria-expanded="true" data-headlessui-state="open" class="QA__button" data-test-id="qa-button-isMyDataSafe" aria-controls="headlessui-disclosure-panel-v-11"><span data-v-bcc04aab="">Is my data safe?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0 rotate-180 transform" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <div data-v-bcc04aab="" id="headlessui-disclosure-panel-v-11" data-headlessui-state="open" class="QA__panel" data-test-id="qa-panel-isMyDataSafe">
        <div data-v-bcc04aab="">The data is stored in your browser's local storage and is not shared with any third parties. You can at any time you wish reset the app using the button below.</div>
      </div>
    </transition-stub><button data-v-bcc04aab="" id="headlessui-disclosure-button-v-12" type="button" aria-expanded="true" data-headlessui-state="open" class="QA__button" data-test-id="qa-button-contact" aria-controls="headlessui-disclosure-panel-v-13"><span data-v-bcc04aab="">How to get in touch?</span><svg data-v-bcc04aab="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0 rotate-180 transform" viewBox="0 0 24 24" aria-hidden="true">
        <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z" class="transition-all"></path>
      </svg></button>
    <transition-stub data-v-bcc04aab="" enteractiveclass="transition duration-100 ease-out" enterfromclass="transform h-0 opacity-0" entertoclass="transform h-full opacity-100" leaveactiveclass="transition duration-75 ease-out" leavefromclass="transform h-full opacity-100" leavetoclass="transform h-0 opacity-0" appear="false" persisted="false" css="true">
      <div data-v-bcc04aab="" id="headlessui-disclosure-panel-v-13" data-headlessui-state="open" class="QA__panel" data-test-id="qa-panel-contact">
        <div data-v-bcc04aab="">You can find the project page in <a data-v-bcc04aab="" href="https://github.com/ajuvonen/veggies/" target="_blank" noreferrer="" noopener="">GitHub</a>. The page has separate boards for reporting issues and engaging in other discussion. Thank you for your interest towards the project!</div>
      </div>
    </transition-stub>
  </div>
</div>"
`;
